cmd/kubectl-gadget/seccompadvisor.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+
+	"github.com/kinvolk/inspektor-gadget/cmd/kubectl-gadget/utils"
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/api/v1alpha1"
+)
+
+var seccompAdvisorCmd = &cobra.Command{
+	Use:   "seccomp-advisor",
+	Short: "Generate seccomp policies based on recorded syscalls activity",
+}
+
+var seccompAdvisorMonitorCmd = &cobra.Command{
+	Use:   "start",
+	Short: "Start to monitor the system calls",
+	Run:   runSeccompAdvisorStart,
+}
+
+var seccompAdvisorReportCmd = &cobra.Command{
+	Use:   "stop",
+	Short: "Stop monitoring and report the policies",
+	Run:   runSeccompAdvisorStop,
+}
+
+var seccompAdvisorListCmd = &cobra.Command{
+	Use:   "list",
+	Short: "List existing seccomp traces",
+	Run:   runSeccompAdvisorList,
+}
+
+var (
+	seccompProfile string
+)
+
+func init() {
+	// Add generic information.
+	rootCmd.AddCommand(seccompAdvisorCmd)
+	utils.AddCommonFlags(seccompAdvisorCmd, &params)
+
+	seccompAdvisorCmd.AddCommand(seccompAdvisorMonitorCmd)
+	seccompAdvisorMonitorCmd.PersistentFlags().StringVarP(&seccompProfile, "seccomp-profile", "", "", "ExternalResource name")
+
+	seccompAdvisorCmd.AddCommand(seccompAdvisorReportCmd)
+	seccompAdvisorCmd.AddCommand(seccompAdvisorListCmd)
+}
+
+// runSeccompAdvisorStart starts monitoring of syscalls for the given
+// parameters.
+func runSeccompAdvisorStart(cmd *cobra.Command, args []string) {
+	if params.Podname == "" {
+		fmt.Fprintf(os.Stderr, "Usage: kubectl gadget seccomp-advisor monitor -p podname\n")
+
+		os.Exit(1)
+	}
+
+	traceOutput := ""
+	traceOutputMode := "Status"
+
+	if seccompProfile != "" {
+		traceOutput = seccompProfile
+		traceOutputMode = "ExternalResource"
+	}
```

Comment:
I am not sure I like this. I think I would prefer to have an additional flag that specifies the output mode.

@mauriciovasquezbernal what do you think?

Reply:
mmm, yes, I think you're right. There should be an output mode and `seccompProfile` will be only meaningful if the output mode is `seccomp-profile`.

Reply:
Mmmm `seccomp-profile`? I am sorry but I am not sure I am following you. 

I don't have clear yet what should be the possible values of an output mode flag but I am not sure which of them `seccomp-profile` identifies. Should it be use for printing the policies directly on the terminal or to apply them with the operator? Or what did I miss?

Reply:
I am fine with adding such an option, but I have two interrogations:
1. What will be the output when using `Stream`? Will the whole seccomp policy printed again each time a new syscall is encountered?
2. How should we deal with it on command line? I mean we will have one option `--outputmode=[status|stream|externalresource|file]` with another `--output=` which will be valid only when `outputmode` is `externalresourece` or `file`?

Reply:
$Currently two modes are supported by the seccomp advisor: (1) storing the policy in the output status and (2) creating a secomp profile for the seccomp operator. 

Right now this PR handles those two modes by having a  `seccomp-profile` parameter, if this is set then the seccomp advisor creates a seccomp profile with the name passed in that parameter. If it's not set then it prints the policy to the terminal. 

My idea is to have an additional parameter that tells the mode to use, `terminal` vs `seccomp-profile` or something like that. 

```
# print to the terminal
$. /kubectl-gadget seccomp-advisor start -p foo -mode terminal

# print to the terminal (by default)
$ ./kubectl-gadget seccomp-advisor start -p foo

# create a bar seccomp profile
$  /kubectl-gadget seccomp-advisor start -p foo -mode seccomp-profile -secomp-profile-name bar 
```

Does it make sense?

Reply:
> Does it make sense?
>
Absolutely, it makes perfect sense. I am just not sure about the names 🤔 . Let me think about it.

> Currently two modes are supported by the seccomp advisor: (1) storing the policy in the output status and (2) creating a secomp profile for the seccomp operator.
>
Actually, there is a third mode which is `Stream`. It does not support the `generate` operation yet (In the TO-DO list) but similarly to `ExternalResource`, it does support the automatic policy generation upon pod termination. For such generation, we do not need to specify the pod, we could just specify a namespace or labels and the gadget will print (in case of `Stream`) or apply (in case of `ExternalResource`) the policy of the pods that terminate while the gadget is running. Notice that for this kind of generation, we do not need to annotate the trace with `generate`.

In a future PR, once we will support also the `generate` operation for the `Stream` output mode and we will have implemented a couple of necessary enhancements ([#265](https://github.com/kinvolk/inspektor-gadget/issues/265), [#377](https://github.com/kinvolk/inspektor-gadget/issues/377) [#379](https://github.com/kinvolk/inspektor-gadget/issues/379)), we could use it with the `--attach` flag:
```
# It prints the trace-id and creates the trace with Trace.OutputMode=Status
$. /kubectl-gadget seccomp-advisor start -p foo -mode terminal
sTL0lkzNctFSQqpb

Then, we will have to stop the gadget to get all the policies.
$. /kubectl-gadget seccomp-advisor stop sTL0lkzNctFSQqpb
... Policies for the pods that terminated and for the pods that are still running ...
```

```
# It blocks the terminal and creates the trace with Trace.OutputMode=Stream. By doing so, we will be able to show in "real-time" the profiles of the pods that terminate.
$. /kubectl-gadget seccomp-advisor start -p foo -mode terminal --attach
Tracing system calls ... Hit Ctrl+C to generate the policies

... Policies for every pod that terminates ...

# Then, when we hit Cltr+C, it will generate the profiles of the pods that are still running.
^C

... Policies for the pods that are still running ...
```

Reply:
> Actually, there is a third mode which is Stream

I have completely forgotten about it. 

> In a future PR, once we will support also the generate operation for the Stream output mode and we will have implemented a couple of necessary enhancements (#265, #377 #379), we could use it with the --attach flag:

Yes, I agree. At that time we could discuss the UI in detail.